### PR TITLE
Improve docs related to version ordering and conflict resolution

### LINKED ISSUE #20329: Dependency resolution depends on existence of external POMs
Consider something like:
```
dependencies {
    implementation 'org.jetbrains.kotlin:kotlin-stdlib:1.4.30-999'
    implementation 'org.jetbrains.kotlin:kotlin-stdlib:1.4.30'
}
```

(In practice, these dependency versions might come transitively)

### Expected Behavior
The build fails. Resolution selects `1.4.30-999` because it is newer than `1.4.30`, but `1.4.30-999` does not exist in any repository.

### Current Behavior
The build uses the `1.4.30` version of the Kotlin stdlib. Gradles tries to find `1.4.30-999`, notices it doesn't exist, and falls back to the older version.

### Context
I originally noticed this because we are depending on `org.jetbrains.kotlinx:kotlinx-serialization-json:1.1.0-RC`, which [depends on](https://mvnrepository.com/artifact/org.jetbrains.kotlinx/kotlinx-serialization-json-jvm/1.1.0-RC) `org.jetbrains.kotlin:kotlin-stdlib:1.4.30-270` (an released version of the kotlin stdlib), and our builds did not fail as I expected they would.

I believe the current behavior can break hermeticity in confusing ways. For example, imagine:
1. `A` depends on `B` and `C`. `B` depends on `C`.
2. The author of `B` publishes a `B'` that uses a new version of `C`, say `C'`. However, they forget to publish `C'`. 
3. `A` upgrades to `B'`. At this point, `A`'s builds will continue to use `C`.
4. The library author publishes `C'`. `A`'s builds will silently start using `C'` (and possibly break).

I think it's reasonable to assume mutations / deletions of artifacts may break builds. However, I'd expect publishing a new artifact to be safe, as long as the build is not using dynamic versions.

### Steps to Reproduce 
See above.

### Your Environment
Gradle 7.2

### PR DESCRIPTION
**Pull request by jbartok**
Fixes #20329

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/dependency_resolution.adoc [hunk 1]
```diff
@@ -85,6 +86,55 @@ As you have seen, Gradle supports a concept of <<rich_versions.adoc#rich-version
...
+|b1-2-3.3
+|b
+|2-3.3
+|===
+
+As you can see separators are any of the `.`, `-`, `_`, `+` characters, plus the empty string when a numeric and a non-numeric part of the version are next to each-other.
+
+When resolving the conflict between competing versions, the following logic applies:
+* first the versions with the highest base version are selected, the rest are discarded
+* if there are still multiple competing versions left, then one is picked with a preference for not having a qualifier or having release status.
```
Comment (jbartok) [inline subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/dependency_resolution.adoc:136]: I'm pondering if we should explain what "release status" is, but a) I myself have no clue (I'm assuming it's not something very prevalent in typical use-cases), and b) it's already getting confusing enough.

**Review by ljacomet — CHANGES_REQUESTED** [review]
Thanks for attempting to improve this. See below for comments.

File: subprojects/docs/src/docs/userguide/dep-man/02-declaring-dependency-versions/single_versions.adoc [hunk 2]
```diff
@@ -35,15 +36,13 @@ Versions are ordered based on the following rules:
...
-** A version with an extra numeric part is considered **higher** than a version without: `1.1` < `1.1.0`
+** If both are non-numeric, the parts are compared **alphabetically**, in a **case-sensitive** manner: `1.A` < `1.B` < `1.a` < `1.b`
+** A version with an extra numeric part is considered **higher** than a version without (even when it's zero): `1.1` < `1.1.0`
 ** A version with an extra non-numeric part is considered **lower** than a version without: `1.1.a` < `1.1`
-* Certain string values have special meaning for the purposes of ordering:
-** The string `dev` is consider **lower** than any other string part: `1.0-dev` < `1.0-alpha` < `1.0-rc`.
-** The strings `rc`, `final`, `ga` and `release` are considered **higher** than any other string part (sorted in that order): `1.0-zeta` < `1.0-RC` < `1.0-FINAL` < `1.0-GA` < `1.0-RELEASE` < `1.0`.
-** The string `SP` will be ordered higher than `release`, it remains however lower than an unqualified version, limiting its use to versioning schemes using either `FINAL`, `GA` or `RELEASE`: `1.0-RELEASE` < `1.0-SP1` < `1.0`
```
[the remark this answers is not in the document]
Comment (jbartok) [reply subprojects/docs/src/docs/userguide/dep-man/02-declaring-dependency-versions/single_versions.adoc:44]: The description was overly complicated and confusing. While `dev` is indeed different than the rest, the other special qualifiers are all the same, besides having an ordering defined for them. So `sp` is special only in that it's last in the defined ordering. Mentioning it separately in the documentation, in my opinion, only makes it harder to understand.

Here, in the GH diff it looks weird, much better in IntelliJ:
image

**Review by ljacomet — APPROVED** [review]

**Comment by ljacomet** [comment]
@jbartok This is good to go but needs to be rebased on `release` for 7.5

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Pull request merged**